### PR TITLE
[MLIR:LITE] Verify unpack op

### PR DESCRIPTION
**Pull request by WindQAQ**
Please inform me if any check is missing :-)

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/unpack.cc#L34-L83

### CODE REVIEW

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc [hunk 1]
```diff
@@ -1444,12 +1444,50 @@ void FakeQuantOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
...
+  int64_t axis_value = op.axis().getSExtValue();
+  if (axis_value < 0)
+    axis_value += rank;
+  if (axis_value < 0 || axis_value >= rank)
+    return op.emitOpError()
+            << "op attribute 'axis' should be in range [-rank, rank), "
+            << "got rank = " << rank
+            << ", and axis = " << op.axis().getSExtValue();
+
+  llvm::SmallVector<int64_t, 4> output_shape;
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:1475]: Could you extract these into a type op interface's build method? (https://github.com/tensorflow/tensorflow/blob/001ec7efbed18e9581e859513c5acc76e5aabbe9/tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc#L905) It results in generating more convenient build method and then also verifies the return type (https://github.com/tensorflow/tensorflow/commit/7779f4bbe0f50a77084561fe4308e27f45ec7a2b is an example but has other parts there).

If you can't get that to work, just add a TODO for me here.

**Review by jpienaar — DISMISSED** [review]
Nice, thanks

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc (hunk 1 above)
Comment (WindQAQ) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:1475]: Thanks for pointing out this! It's very powerful. But I have a quick question here. If input is unranked, then how do we infer return type? For `UnpackOp`, the output is variadic, we cannot even know the number of results when the input is unranked. Is there any trick that we can achieve it? Thanks.

Comment (jpienaar) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:1475]: Good question. So the number of outputs are always known (https://github.com/tensorflow/tensorflow/blob/acbfab2b0191ec7a845eb967b39f1cb08d3a3c3a/tensorflow/core/ops/array_ops.cc#L445) their shapes may not be (well, would not be in unranked case). I believe we keep that attribute still, else we would not be able to :-)

Comment (WindQAQ) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.cc]: Thank you for the clarification 😃 Updated.

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc [hunk 2]
```diff
@@ -1443,12 +1443,66 @@ void FakeQuantOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
...
 // TODO(b/133486129): Implement shape inference for unpack
 
-static LogicalResult Verify(UnpackOp op) {
-  // TODO(antiagainst): Implement other checks as in
-  // tensorflow/lite/kernels/unpack.cc
+LogicalResult UnpackOp::inferReturnTypes(
+    MLIRContext *context, Optional<Location> location, ValueRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<Type> &inferredReturnTypes) {
+  auto num = attributes.get("num");
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:1450]: I should have mentioned op adaptors here (e.g., you can do UnpackOp::Adaptor op(operands, attributes); and then use its verify method and named accessors). I need to fix up another one too, so just leave "TODO(jpienaar): Use Adaptor instead" here.

**Review by jpienaar — DISMISSED** [review]
Nice, thanks!

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.cc (hunk 2 above)
Comment (WindQAQ) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:1451]: Done :-)

**Review by jpienaar — DISMISSED** [review]

**Review by jpienaar — APPROVED** [review]

**Pull request merged**